Sources/GRPC/Docs.docc/Proposals/0001-stub-api.md

```diff
@@ -0,0 +1,1100 @@
+
+    /// The operation was cancelled (typically by the caller).
+    public static let cancelled = Self(code: 1)
+
+    /// Unknown error. An example of where this error may be returned is if a
+    /// status value received from another address space belongs to an error-space
+    /// that is not known in this address space. Also errors raised by APIs that
+    /// do not return enough error information may be converted to this error.
+    public static let unknown = Self(code: 2)
+
+    // etc.
+  }
+
+  /// The error code.
+  public var code: Code
+
+  /// A message describing the error.
+  public var message: String
+
+  /// Metadata associated with the error.
+  public var metadata: Metadata
+
+  public init(code: Code, message: String, metadata: Metadata = [:]) {
+    // ...
+  }
+}
+```
+
+### Generated server code
+
+Code generated for each service includes two protocols. The first, higher-level
+protocol which most users interact with, includes one function per defined
+method. The shape of the function matches the method definition. For example
+unary methods accept a `ServerRequest.Single` and return a
+`ServerResponse.Single`, bidirectional streaming methods accept a
+`ServerRequest.Stream` and return a `ServerResponse.Stream`.
+
+The second, base protocol, defines each method in terms of streaming requests
+and responses. The higher-level protocol refines the base protocol and provides
+default implementations of methods in the base protocol in terms of their
+higher-level counterpart.
+
+The base protocol is an escape hatch allowing advanced users to have further
+control over their RPCs. As an example, if a service owner needs to respond to
+initial metadata in a client streaming RPC before processing the complete stream
+of messages from the request they could implement their RPC in terms of the
+fully streaming version provided by the base protocol.
+
+Users can throw any error from each method. Since gRPC has a well defined error
+model, gRPC Swift catches errors of type `RPCError` and extracts the code and
+message. The code and message are propagated back to the client as the status of
+the RPC. The library discards all other errors and returns a status with code
+`unknown` to the client.
+
+The following code demonstrates how these protocols would look for the Echo
+service. Some details are elided as they aren't relevant.
+
+```swift
+// (Defined elsewhere.)
+public typealias EchoRequest = ...
+public typealias EchoResponse = ...
+
+/// The generated base protocol for the "Echo" service providing each method
+/// in a fully streamed form.
+///
+/// This protocol should typically not be implemented, instead you should
+/// implement ``EchoServiceProtocol`` which refines this protocol. However, if
+/// you require more granular control over your RPCs then they may implement
+/// this protocol, or methods from this protocol, instead.
+public protocol EchoServiceStreamingProtocol: BindableService, Sendable {
```

Comment:
I find the name `BindableService` interesting. Especially with `Service` from `ServiceLifecycle`. What do you think about naming it `GRPCService`?

Reply:
I'm open to bikeshedding the naming. I believe grpc-java uses `BindableService` but I'm not wedded to that. I would be tempted to drop the "GRPC" and just call it `Service`: all generated code should be fully qualified anyway so would spell it `GRPC.Service` so I don't think ambiguity would be an issue.

Reply:
Agreed we can bikeshed later I think we should not use common types in our libraries. We have made a similar decision in `Kafka`, `Redis` etc. where we use `XXXClient` instead of just `Client`. `Service` is particularly dangerous since `ServiceLifecycle` uses it and it is probably quite common to have both in the same scope.

Since we have `RPCError` maybe just `RPCService`?

Reply:
I don't think `Service` would be an issue for the reasons given above but I'm also not opposed to `RPCService`.